src/msg/async/ProtocolV2.cc

```diff
@@ -1151,9 +1151,12 @@ CtPtr ProtocolV2::read_frame_segment() {
   rx_segments_data.emplace_back();
 
   uint32_t onwire_len = rx_frame_asm.get_segment_onwire_len(seg_idx);
-  uint16_t align = rx_frame_asm.get_segment_align(seg_idx);
+  if (onwire_len == 0) {
+    return _handle_read_frame_segment();
```

Comment:
IIUC `Message` frames dominate typical exchanges. Is there any other reason behind the change than the cold-path optimization?

Reply:
I don't think it's cold.  Pretty much all data-bearing messages have an empty segment: middle.  But even if that wasn't the case, eliminating a useless `buffer::ptr_node` heap allocation while also making the empty segment case explicit (rather than relying on `READ[_RXBUF](0, ...)` to do the right thing) is definitely a win.

Reply:
> Pretty much all data-bearing messages have an empty segment: middle.

OK, then it's on the hot path.

> making the empty segment case explicit

Makes sense.